src/common/walFilter/versions/recordProcessGPDB7.c

```diff
@@ -0,0 +1,461 @@
+            XLR_BLOCK_ID_DATA_SHORT,
+            (uint8_t) (record->xl_tot_len - sizeof(XLogRecordGPDB7) - sizeof(XLogRecordDataHeaderShort))
+        };
+    }
+    else
+    {
+        *((XLogRecordDataHeaderLong *) recordData) = (XLogRecordDataHeaderLong){
+            XLR_BLOCK_ID_DATA_LONG,
+            (uint32_t) (record->xl_tot_len - sizeof(XLogRecordGPDB7) - sizeof(XLogRecordDataHeaderLong))
+        };
+    }
+}
+
+FN_EXTERN void
+filterRecordGPDB7(XLogRecordBase *const recordBase, const PgPageSize pageSize)
+{
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(INT, (int) pageSize);
+    FUNCTION_LOG_END();
+
+    XLogRecordGPDB7 *const record = (XLogRecordGPDB7 *const) recordBase;
+
+    if (record->xl_rmid == RM_XLOG_ID && record->xl_info == XLOG_NOOP)
+        goto end;
+
+    List *const nodes = getRelFileNodes(record, pageSize);
+    if (lstEmpty(nodes)){
+        lstFree(nodes);
+        goto end;
+    }
+
+    bool hasPassFilter = false;
+    bool hasNotPassFilter = false;
+    List *notPassFilterList = lstNewP(sizeof(RelFileNode));
+    for (uint32_t i = 0; i < lstSize(nodes); i++)
+    {
+        const RelFileNode *const relFileNode = lstGet(nodes, i);
+        if (isRelationNeeded(relFileNode->dbNode, relFileNode->spcNode, relFileNode->relNode))
+        {
+            hasPassFilter = true;
+        }
+        else
+        {
+            lstAdd(notPassFilterList, relFileNode);
+            hasNotPassFilter = true;
+        }
+    }
+    lstFree(nodes);
+
+    // For now, I can't find a real-life example of such a case, so let's handle it the safest way.
+    // If a record has two and more relFileNodes and one of them is passing a filter while the others are not, then throw an error.
+    // Since:
+    //  1. If we pass that record, we may get an error during recovery because one of the files will be missing.
+    //  2. If we filter out that record, we can get an inconsistent database.
+    if (hasPassFilter && hasNotPassFilter)
+    {
+        ASSERT(!lstEmpty(notPassFilterList));
+        String *errMessage = strCatZ(strNew(), "The following RefFileNodes cannot be filtered out because they are in the same"
+                                     " XLogRecord as the RefFileNode that passes the filter. [");
+        for (uint32_t i = 0; i < lstSize(notPassFilterList); i++)
+        {
+            if (i > 0)
+                strCatZ(errMessage, ", ");
+            const RelFileNode *const relFileNode = lstGet(notPassFilterList, i);
+            strCatFmt(errMessage, "{%u, %u, %u}", relFileNode->spcNode, relFileNode->dbNode, relFileNode->relNode);
+        }
+        strCatZ(errMessage, "].");
+//        const RelFileNode *const lastRelFileNode = lstGetLast(notPassFilterList);
+//        strCatFmt(errMessage, "{%u, %u, %u}].", lastRelFileNode->spcNode, lastRelFileNode->dbNode, lastRelFileNode->relNode);
+        strCatFmt(errMessage, "\nHINT: Add these RelFileNodes to your filter.");
```

Comment:
Can we use a simpler function - strCatZ? May be one call of strCatZ instead of strCatZ+strCatFmt?